src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,208 @@
+                easy.SetCurlOption(CURLoption.CURLOPT_SSL_CTX_DATA, GCHandle.ToIntPtr(_gcHandle));
+                IgnorePeerCert(easy);
+                VerboseTrace("setting client certificate options");
+            }
+
+            public void Dispose()
+            {
+                _gcHandle.Free();
+                if (_privateKeyHandle != null)
+                {
+                    _privateKeyHandle.Dispose();
+                }
+
+                if (_certHandle != null)
+                {
+                    _certHandle.Dispose();
+                }
+            }
+
+            private int TlsClientCertCallback(IntPtr ssl, out IntPtr certHandle, out IntPtr privateKeyHandle)
+            {
+                using (Interop.libssl.SafeSslHandle sslHandle = new Interop.libssl.SafeSslHandle(ssl))
+                {
+                    certHandle = IntPtr.Zero;
+                    privateKeyHandle = IntPtr.Zero;
+                    VerboseTrace("libssl's client certificate callback");
+                    ISet<string> issuerNames = GetRequestCertificateAuthorities(sslHandle);
+                    X509Certificate2 certificate;
+                    X509Chain chain;
+                    if (!GetClientCertificate(issuerNames, out certificate, out chain))
+                    {
+                        VerboseTrace("no cert or chain");
+                        return 0;
+                    }
+
+                    Debug.Assert(chain != null && chain.ChainElements.Count > 0, " a X509Chain received must not be empty");
+
+                    using (RSAOpenSsl rsa = certificate.GetRSAPrivateKey() as RSAOpenSsl)
+                    using (ECDsaOpenSsl ecdsa = certificate.GetECDsaPrivateKey() as ECDsaOpenSsl)
+                    {
+                        VerboseTrace("rsa = " + rsa);
+                        if (rsa != null)
+                        {
+                            _privateKeyHandle = rsa.DuplicateKeyHandle();
+                        }
+                        else if (ecdsa != null)
+                        {
+                            _privateKeyHandle = ecdsa.DuplicateKeyHandle();
+                        }
+                    }
+
+                    if (_privateKeyHandle == null)
+                    {
+                        VerboseTrace("invalid private key");
+                        return 0;
+                    }
+                    _certHandle = Interop.libcrypto.X509_dup(certificate.Handle);
+
+                    // add extra chain cert
+                    using (SafeSslContextHandle contextHandle = Interop.libssl.SSL_get_SSL_CTX(sslHandle))
+                    {
+                        for (int i = chain.ChainElements.Count - 2; i > 0; i--)
+                        {
+                            VerboseTrace("here " + contextHandle);
+                            SafeX509Handle dupCertHandle = Interop.libcrypto.X509_dup(chain.ChainElements[i].Certificate.Handle);
```

Comment:
Reference counting for `chain.ChainElements[i].Certificate.Handle`:

Start: x
X509_dup: x+1
AddExtraChainCertToSslCtx: x+2
Eventual dispose/finalize of dupCertHandle: x+1
Eventual dispose/finalize of the X509Certificate2: x
Eventual dispose/finalize of the SafeSslHandle that we fed the cert to: x-1 (if 0, resources are finally freed).

Lines 112-117 should be in a using to avoid making the finalizer do the work.